TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/DimensionsMenuImplementation.php

```diff
@@ -0,0 +1,216 @@
+                    $acceptableCombination = [$pinnedDimensionName => $allDimensionPresets[$pinnedDimensionName]['presets'][$allowedPresetIdentifier]['values']];
+                    $allowedAdditionalPresets = $this->configurationContentDimensionPresetSource->getAllowedDimensionPresetsAccordingToPreselection('country', [$pinnedDimensionName => $allowedPresetIdentifier]);
+                    foreach ($allowedAdditionalPresets as $allowedAdditionalDimensionName => $allowedAdditionalPreset) {
+                        $acceptableCombination[$allowedAdditionalDimensionName] = $allowedAdditionalPreset['presets'][$allowedAdditionalPreset['defaultPreset']]['values'];
+                    }
+                    $nodeInDimensions = $this->getNodeInDimensions($acceptableCombination, []);
+                    if ($nodeInDimensions !== null) {
+                        continue;
+                    }
+                }
+            }
+
+            if ($nodeInDimensions !== null && $this->isNodeHidden($nodeInDimensions)) {
+                $nodeInDimensions = null;
+            }
+
+            // determine metadata for target dimensions of node
+            array_walk($targetDimensions, function (&$dimensionValue, $dimensionName, $allDimensionPresets) use ($pinnedDimensionName) {
+                $dimensionValue = [
+                    'value' => $dimensionValue,
+                    'label' => $allDimensionPresets[$dimensionName]['presets'][$dimensionValue]['label'],
+                    'isPinnedDimension' => ($pinnedDimensionName === null || $dimensionName == $pinnedDimensionName) ? true : false
+                ];
+            }, $allDimensionPresets);
+
+            if ($pinnedDimensionName === null) {
+                $itemLabel = Utility::evaluateEelExpression($this->getLabelExpression(), $this->eelEvaluator, array('node' => $nodeInDimensions), $this->defaultContextConfiguration);
+            } else {
+                $itemLabel = $targetDimensions[$pinnedDimensionName]['label'];
+            }
+
+            $menuItems[] = [
+                'node' => $nodeInDimensions,
+                'state' => $this->calculateItemState($nodeInDimensions),
+                'label' => $itemLabel,
+                'dimensions' => $allowedCombination,
+                'targetDimensions' => $targetDimensions,
+                'pinnedDimensionName' => $pinnedDimensionName
+            ];
+        }
+
+        // sort/limit according to configured "presets" if needed
+        if ($pinnedDimensionName !== null && is_array($pinnedDimensionValues)) {
+            $sortedMenuItems = [];
+            foreach ($pinnedDimensionValues as $pinnedDimensionValue) {
+                foreach ($menuItems as $menuItemKey => $menuItem) {
+                    if ($menuItem['targetDimensions'][$pinnedDimensionName]['value'] === $pinnedDimensionValue) {
+                        $sortedMenuItems[$menuItemKey] = $menuItem;
+                    }
+                }
+            }
+
```

Comment:
Can't this be done with some kind of usort that compares the index value of the $pinnedDimensionValue?
I think that would be more explicit about what's going on and also a bit less complex. I can give it a try.